File: src-tauri/src/printer.rs
use serde::Deserialize;
use std::io::Write;
use std::net::{IpAddr, TcpStream};
use std::time::Duration;
use tauri::State;

use crate::{auth::require_auth, session::SessionState};

/// Validate printer host to prevent SSRF attacks.
/// Only allows private/local network IPs and simple hostnames.
fn validate_printer_host(host: &str) -> Result<(), String> {
    if host.is_empty() {
        return Err("IP/host printer wajib diisi".into());
    }
    // Block scheme injection (http://, https://, file://, etc.)
    if host.contains("://") || host.contains('/') || host.contains('@') {
        return Err("Format host printer tidak valid".into());
    }
    // If it's an IP address, validate it's a private/local address
    if let Ok(ip) = host.parse::<IpAddr>() {
        match ip {
            IpAddr::V4(v4) => {
                if v4.is_loopback() || v4.is_private() || v4.is_link_local() {
                    return Ok(());
                }
                return Err("Printer harus berada di jaringan lokal (private IP)".into());
            }
            IpAddr::V6(v6) => {
                if v6.is_loopback() {
                    return Ok(());
                }
                return Err("Printer harus berada di jaringan lokal (IPv4 private)".into());
            }
        }
    }
    // Simple hostname validation — allow alphanumeric, dots, hyphens
    if host
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '.' || c == '-')
        && !host.starts_with('-')
        && !host.ends_with('-')
    {
        return Ok(());
    }
    Err("Format host printer tidak valid".into())
}

#[derive(Debug, Deserialize)]
pub struct PrintReceiptItemPayload {
    pub name: String,
    pub quantity: i64,
    pub unit_price: f64,
    pub subtotal: f64,
}

#[derive(Debug, Deserialize)]
pub struct PrintReceiptPayload {
    pub host: String,
    pub port: Option<u16>,
    pub store_name: Option<String>,
    pub invoice_no: String,
    pub payment_method: String,
    pub total_amount: f64,
    pub cash_received: Option<f64>,
    pub change_amount: Option<f64>,
    pub items: Vec<PrintReceiptItemPayload>,
}

fn escpos_line(out: &mut Vec<u8>, text: &str) {
    out.extend_from_slice(text.as_bytes());
    out.push(b'\n');
}

#[tauri::command]
pub fn print_thermal_receipt(
    session: State<'_, SessionState>,
    payload: PrintReceiptPayload,
) -> Result<bool, String> {
    let _user = require_auth(&session)?;
    let host = payload.host.trim();
    validate_printer_host(host)?;
    let port = payload.port.unwrap_or(9100);
    let mut bytes = Vec::new();
    bytes.extend_from_slice(&[0x1b, 0x40]);
    bytes.extend_from_slice(&[0x1b, 0x61, 0x01]);
    bytes.extend_from_slice(&[0x1d, 0x21, 0x11]);
    escpos_line(
        &mut bytes,
        payload.store_name.as_deref().unwrap_or("CatatAgen Local"),
    );
    bytes.extend_from_slice(&[0x1d, 0x21, 0x00]);
    escpos_line(&mut bytes, &payload.invoice_no);
    escpos_line(&mut bytes, "------------------------------");
    bytes.extend_from_slice(&[0x1b, 0x61, 0x00]);
    for item in payload.items {
        escpos_line(&mut bytes, &item.name.chars().take(30).collect::<String>());
        escpos_line(
            &mut bytes,
            &format!(
                "{} x Rp{:.0} = Rp{:.0}",
                item.quantity, item.unit_price, item.subtotal
            ),
        );
    }
    escpos_line(&mut bytes, "------------------------------");
    escpos_line(&mut bytes, &format!("Bayar: {}", payload.payment_method));
    escpos_line(&mut bytes, &format!("TOTAL: Rp{:.0}", payload.total_amount));
    if let Some(cash_received) = payload.cash_received {
        escpos_line(&mut bytes, &format!("Tunai: Rp{:.0}", cash_received));
    }
    if let Some(change_amount) = payload.change_amount {
        escpos_line(&mut bytes, &format!("Kembali: Rp{:.0}", change_amount));
    }
    bytes.extend_from_slice(&[0x1b, 0x61, 0x01]);
    escpos_line(&mut bytes, "Terima kasih");
    bytes.extend_from_slice(b"\n\n\n");
    bytes.extend_from_slice(&[0x1d, 0x56, 0x42, 0x00]);
    let mut stream = TcpStream::connect((host, port))
        .map_err(|e| format!("Gagal konek printer {host}:{port}: {e}"))?;
    stream.set_write_timeout(Some(Duration::from_secs(5))).ok();
    stream
        .write_all(&bytes)
        .map_err(|e| format!("Gagal kirim struk ke printer: {e}"))?;
    Ok(true)
}

File: src/routes/printer.js
/**
 * Printer route — send ESC/POS thermal receipt via TCP.
 */

import { Router } from 'express';
import net from 'node:net';
import { requireAuth } from '../middleware/auth.js';

const router = Router();

function validatePrinterHost(host) {
  if (!host || !host.trim()) throw new Error('IP/host printer wajib diisi');
  host = host.trim();
  if (host.includes('://') || host.includes('/') || host.includes('@')) {
    throw new Error('Format host printer tidak valid');
  }
  // IP validation
  if (/^\d+\.\d+\.\d+\.\d+$/.test(host)) {
    const parts = host.split('.').map(Number);
    if (parts[0] === 127 || (parts[0] === 10) || (parts[0] === 172 && parts[1] >= 16 && parts[1] <= 31) || (parts[0] === 192 && parts[1] === 168)) {
      return host;
    }
    throw new Error('Printer harus berada di jaringan lokal (private IP)');
  }
  if (/^[a-zA-Z0-9.-]+$/.test(host) && !host.startsWith('-') && !host.endsWith('-')) {
    return host;
  }
  throw new Error('Format host printer tidak valid');
}

function escposLine(out, text) {
  out.extend(Buffer.from(text, 'utf-8'));
  out.push(0x0a); // \n
}

// POST /api/printer/receipt
router.post('/printer/receipt', requireAuth, (req, res) => {
  try {
    const host = validatePrinterHost(req.body.host || '');
    const port = Number(req.body.port) || 9100;
    const {
      store_name, invoice_no, payment_method, total_amount,
      cash_received, change_amount, items,
    } = req.body;

    const bytes = [];

    // ESC/POS init
    bytes.push(0x1b, 0x40); // Initialize
    bytes.push(0x1b, 0x61, 0x01); // Center align
    bytes.push(0x1d, 0x21, 0x11); // Double height, double width

    escposLine(bytes, store_name || 'CatatAgen Local');

    bytes.push(0x1d, 0x21, 0x00); // Normal size
    escposLine(bytes, invoice_no || '');
    escposLine(bytes, '------------------------------');

    bytes.push(0x1b, 0x61, 0x00); // Left align

    for (const item of (items || [])) {
      escposLine(bytes, (item.name || '').substring(0, 30));
      escposLine(bytes, `${item.quantity} x Rp${Math.round(item.unit_price || 0)} = Rp${Math.round(item.subtotal || 0)}`);
    }

    escposLine(bytes, '------------------------------');
    escposLine(bytes, `Bayar: ${payment_method || 'cash'}`);
    escposLine(bytes, `TOTAL: Rp${Math.round(total_amount || 0)}`);

    if (cash_received) {
      escposLine(bytes, `Tunai: Rp${Math.round(cash_received)}`);
    }
    if (change_amount) {
      escposLine(bytes, `Kembali: Rp${Math.round(change_amount)}`);
    }

    bytes.push(0x1b, 0x61, 0x01); // Center
    escposLine(bytes, 'Terima kasih');
    bytes.push(0x0a, 0x0a, 0x0a);
    bytes.push(0x1d, 0x56, 0x42, 0x00); // Cut paper

    const buffer = Buffer.from(bytes);

    // Send via TCP
    const socket = new net.Socket();
    socket.setTimeout(5000);

    socket.connect(port, host, () => {
      socket.write(buffer, (err) => {
        socket.destroy();
        if (err) {
          return res.status(500).json({ error: `Gagal kirim struk ke printer: ${err.message}` });
        }
        res.json({ ok: true });
      });
    });

    socket.on('error', (err) => {
      socket.destroy();
      res.status(500).json({ error: `Gagal konek printer ${host}:${port}: ${err.message}` });
    });

    socket.on('timeout', () => {
      socket.destroy();
      res.status(500).json({ error: `Timeout koneksi printer ${host}:${port}` });
    });
  } catch (err) {
    res.status(400).json({ error: err.message });
  }
});

export default router;
